src/Analysis/Ast/Impl/Types/PythonClassType.Generics.cs

```diff
@@ -0,0 +1,358 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Python.Analysis.Specializations.Typing;
+using Microsoft.Python.Analysis.Utilities;
+using Microsoft.Python.Analysis.Values;
+using Microsoft.Python.Core;
+
+namespace Microsoft.Python.Analysis.Types {
+    internal partial class PythonClassType : PythonType, IPythonClassType, IGenericType, IEquatable<IPythonClassType> {
+        private Dictionary<string, PythonClassType> _specificTypeCache;
+        private Dictionary<IGenericTypeParameter, IPythonType> _genericParameters;
+        private ReentrancyGuard<IPythonClassType> _genericSpecializationGuard = new ReentrancyGuard<IPythonClassType>();
+        private ReentrancyGuard<IPythonClassType> _genericResolutionGuard = new ReentrancyGuard<IPythonClassType>();
+
+        #region IGenericType
+        /// <summary>
+        /// List of unfilled generic type parameters. Represented as entries in the GenericParameters dictionary 
+        /// that have both key and value as generic type parameters
+        /// e.g
+        /// {T, T}
+        /// Where T is a generic type parameter that needs to be filled in by the class
+        /// </summary>
+        public virtual IReadOnlyList<IGenericTypeParameter> Parameters => GenericParameters.Values.
+            Distinct().OfType<IGenericTypeParameter>().ToList();
+
+        /// <summary>
+        /// A Class is generic if it has one or more unfilled generic type parameters or one of its bases is generic
+        /// </summary>
+        public virtual bool IsGeneric {
+            get {
+                if (!_genericResolutionGuard.Push(this)) {
+                    return false;
+                }
+
+                try {
+                    return !Parameters.IsNullOrEmpty() || (Bases?.OfType<IGenericType>().Any(g => g.IsGeneric) ?? false);
+                } finally {
+                    _genericResolutionGuard.Pop();
+                }
+            }
+        }
+
+        #endregion
+
+
+        /// <summary>
+        /// Gets a list of distinct type parameters from bases and the class itself
+        /// </summary>
+        private IGenericTypeParameter[] GetTypeParameters() {
```

Comment:
See threading comment above.

Reply:
This method only reads so it should be fine.